daemon/graphdriver/zfs/zfs.go

```diff
@@ -0,0 +1,277 @@
+}
+
+type Logger struct{}
+
+func (*Logger) Log(cmd []string) {
+	log.Debugf("[zfs] %s", strings.Join(cmd, " "))
+}
+
+func Init(base string, opt []string) (graphdriver.Driver, error) {
+	var err error
+	options, err := parseOptions(opt)
+	options.mountPath = base
+	if err != nil {
+		return nil, err
+	}
+
+	rootdir := path.Dir(base)
+
+	if options.fsName == "" {
+		err = checkRootdirFs(rootdir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if _, err := exec.LookPath("zfs"); err != nil {
+		return nil, fmt.Errorf("zfs command is not available")
+	}
+
+	file, err := os.OpenFile("/dev/zfs", os.O_RDWR, 600)
+	defer file.Close()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to initialize: %v", err)
+	}
+
+	if options.fsName == "" {
+		options.fsName, err = lookupZfsDataset(rootdir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	logger := Logger{}
+	zfs.SetLogger(&logger)
+
+	dataset, err := zfs.GetDataset(options.fsName)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot open %s", options.fsName)
+	}
+
+	d := &Driver{
+		dataset: dataset,
+		options: options,
+	}
+	return graphdriver.NaiveDiffDriver(d), nil
+}
+
+func parseOptions(opt []string) (ZfsOptions, error) {
+	var options ZfsOptions
+	options.fsName = ""
+	for _, option := range opt {
+		key, val, err := parsers.ParseKeyValueOpt(option)
+		if err != nil {
+			return options, err
+		}
+		key = strings.ToLower(key)
+		switch key {
+		case "zfs.fsname":
+			options.fsName = val
+		default:
+			return options, fmt.Errorf("Unknown option %s\n", key)
+		}
+	}
+	return options, nil
+}
+
+func checkRootdirFs(rootdir string) error {
+	var buf syscall.Statfs_t
+	if err := syscall.Statfs(rootdir, &buf); err != nil {
+		fmt.Errorf("Failed to access '%s': %s", rootdir, err)
+	}
+
+	if graphdriver.FsMagic(buf.Type) != graphdriver.FsMagicZfs {
+		log.Debugf("[zfs] no zfs dataset found for rootdir '%s'", rootdir)
+		return graphdriver.ErrPrerequisites
+	}
+	return nil
+}
+
+var CprocMounts = C.CString("/proc/mounts")
+var CopenMod = C.CString("r")
+
+func lookupZfsDataset(rootdir string) (string, error) {
+	var stat syscall.Stat_t
+	var Cmnt C.struct_mntent
+	var Cfp *C.FILE
+	buf := string(make([]byte, 256, 256))
+	Cbuf := C.CString(buf)
+	defer free(Cbuf)
+
+	if err := syscall.Stat(rootdir, &stat); err != nil {
+		return "", fmt.Errorf("Failed to access '%s': %s", rootdir, err)
+	}
+	wantedDev := stat.Dev
+
+	if Cfp = C.setmntent(CprocMounts, CopenMod); Cfp == nil {
+		return "", fmt.Errorf("Failed to open /proc/mounts")
+	}
+	defer C.endmntent(Cfp)
+
+	for C.getmntent_r(Cfp, &Cmnt, Cbuf, 256) != nil {
+		dir := C.GoString(Cmnt.mnt_dir)
+		if err := syscall.Stat(dir, &stat); err != nil {
+			log.Debugf("[zfs] failed to stat '%s' while scanning for zfs mount: %v", dir, err)
+			continue // may fail on fuse file systems
+		}
+
+		fs := C.GoString(Cmnt.mnt_type)
+		if stat.Dev == wantedDev && fs == "zfs" {
+			return C.GoString(Cmnt.mnt_fsname), nil
+		}
+	}
+	// should never happen
```

Comment:
I think, the zfs mount could disappear between statfs and reading /proc/mounts or beween reading /proc/mounts and the Stats syscalls.
So returning an error should be ok here:
1. If the user has explicitly request the zfs driver, the deamon will be stopped with this error message
2. otherwise the next graphdriver will be tried.